File: src/p0016/java/solution.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        int minSum = 0, minDelta = Integer.MAX_VALUE;
        Arrays.sort(nums);
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1, k = nums.length - 1; j < k; ) {
                int sum = nums[i] + nums[j] + nums[k];
                if (sum < target) {
                    j++;
                } else if (sum > target) {
                    k--;
                } else {
                    return sum;
                }
                int delta = Math.abs(sum - target);
                if (minDelta > delta) {
                    minSum = sum;
                    minDelta = delta;
                }
            }
        }
        return minSum;
    }
}

File: src/p0016/golang/solution.go
func threeSumClosest(nums []int, target int) int {
	minSum, minDelta := 0, math.MaxInt32
	sort.Ints(nums)
	for i := 0; i < len(nums); i++ {
		for j, k := i+1, len(nums)-1; j < k; {
			sum := nums[i] + nums[j] + nums[k]
			if sum < target {
				j++
			} else if sum > target {
				k--
			} else {
				return sum
			}
			delta := sum - target
			if delta < 0 {
				delta = -delta
			}
			if minDelta > delta {
				minSum, minDelta = sum, delta
			}
		}
	}
	return minSum
}
